std/net/curl.d

```diff
@@ -3565,6 +3568,102 @@ import std.typecons : Flag;
 /// Flag to specify whether or not an exception is thrown on error.
 alias ThrowOnError = Flag!"throwOnError";
 
+private struct CurlAPI
+{
+    static struct API
+    {
+    extern(C):
+        import core.stdc.config : c_long;
+        CURLcode function(c_long flags) global_init;
+        void function() global_cleanup;
+        curl_version_info_data * function(CURLversion) version_info;
+        CURL* function() easy_init;
+        CURLcode function(CURL *curl, CURLoption option,...) easy_setopt;
+        CURLcode function(CURL *curl) easy_perform;
+        CURL* function(CURL *curl) easy_duphandle;
+        char* function(CURLcode) easy_strerror;
+        CURLcode function(CURL *handle, int bitmask) easy_pause;
+        void function(CURL *curl) easy_cleanup;
+        curl_slist* function(curl_slist *, char *) slist_append;
+        void function(curl_slist *) slist_free_all;
+    }
+    __gshared API _api;
+    __gshared void* _handle;
+
+    static ref API instance() @property
+    {
+        import std.concurrency;
+        initOnce!_handle(loadAPI());
+        return _api;
+    }
+
+    static void* loadAPI()
+    {
+        void* handle;
+        version (OSX)
+        {
+            import core.sys.posix.dlfcn;
+            handle = dlopen("libcurl.4.dylib", RTLD_LAZY);
+        }
+        else version (Posix)
+        {
+            import core.sys.posix.dlfcn;
+            static immutable names = ["libcurl.so", "libcurl.so.4", "libcurl-gnutls.so.4", "libcurl.so.3"];
+            foreach (name; names)
+            {
+                handle = dlopen(name.ptr, RTLD_LAZY);
+                if (handle !is null) break;
+            }
+        }
+        else version (Windows)
+        {
+            import core.sys.windows.windows;
+            handle = LoadLibraryA("libcurl.dll");
+        }
+        else
+            static assert(0, "unimplemented");
+
+        enforce!CurlException(handle !is null, "Failed to load libcurl.");
+        foreach (mem; __traits(allMembers, API))
+        {
+            version (Posix)
+                void* p = dlsym(handle, "curl_"~mem);
+            version (Windows)
+                void* p = GetProcAddress(handle, "curl_"~mem);
+
+            __traits(getMember, _api, mem) = cast(typeof(__traits(getMember, _api, mem)))
+                enforce!CurlException(p, "Couldn't load curl_"~mem~" from libcurl.");
```

Comment:
The idea is to try all of them. If there were a chance to fail we'd need to add a way to customize the library name. But it seems that we can avoid this, because the names are pretty standardized.